### PR TITLE
feat: Initial development of seed eval dataset generation

### PR DESCRIPTION
**Pull request by chen0040**
### Description
This PR introduces the initial development of the Seed Evaluation Dataset Generation feature. It provides an autonomous, multi-agent workflow to generate, evaluate, and expand high-quality Natural Language Question (NLQ) and SQL pairs directly from a database schema.

### Key Changes
  * Updated the dataset-generation skill 
  * Added `autoctx:generate-dataset` to `evals/core-cujs/dataset.json` to test the new dataset generation capability in the core CUJs.

### Testing
* Core CUJs updated to include the new dataset generation scenario (`autoctx:generate-dataset*`).

### CODE REVIEW

File: src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/genai_client.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    #         http_status_codes=RETRYABLE_STATUS_CODES,
+    #     ),
+    #     timeout=120 * 1000,
+    # )
+    self._client = genai.Client(
+        api_key=api_key,
+        http_options=types.HttpOptions(
+          retry_options=types.HttpRetryOptions(
+              attempts=1  # 1 attempt means 0 retries
+          ),
```
[the remark this answers is not in the document]
Comment (chen0040) [reply src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/genai_client.py]: It is intended to disable the built-in retry so that the exception can be bubbled up

File: src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/dataset_generator.py [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+from .state_manager import CommittedPairCollection, PendingPairCollection, RejectedPairCollection
+from .state_manager import DatasetGenConfig, DatasetGenDbProfile
+from .state_manager import DatasetGenStates, DatasetGenEventLogger
+from .state_manager import SqlCollection
+
+from .schema import generate_ddls_from_json_schema, SchemaDDL, TableProfile, SchemaProfile, TableSelector
+from .generators import SQLCandidateGenerator, QuestionGenerator, QuestionReviewer, QuestionRefiner
+
+
+assert os.getenv('GEMINI_API_KEY'), "Missing GEMINI_API_KEY environment variable."
```
[the remark this answers is not in the document]
Comment (chen0040) [reply src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/dataset_generator.py]: outdated

File: src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/genai_client.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+          contents=prompt,
+          config=types.GenerateContentConfig(
+              response_mime_type=response_mime_type,
+              response_schema=response_schema,
+              temperature=temperature,
+          ),
+      ), timeout=self.timeout_secs)
+      return result
+    except Exception as e:
+      return None
```
[the remark this answers is not in the document]
Comment (chen0040) [reply src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/genai_client.py]: update based on comments

File: src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/generators.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+      )
+    else:
+      sqls = await self.generate_candidates_4(
+        schema_profile=schema_profile, 
+        dialect=dialect,
+        complexity=complexity,
+        count=count
+      )
+
+    sqls = [sql for sql in sqls if sql and 20 <= len(sql) <= 300]
```
[the remark this answers is not in the document]
Comment (chen0040) [reply src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/generators.py]: update based on comments

File: src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/dataset_generator.py [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+
+        elogger.log_event("refine_nlq_triggered", {
+                "rejected_pairs_sample": rejected_pair_collection.first(),
+                "rejected_pairs_count": len(rejected_pair_collection),
+                "verified_pairs_path": committed_pair_collection.path,
+                "rejected_pairs_path": rejected_pair_collection.path,
+            }, states.to_dict())
+        
+        if not rejected_pair_collection.is_empty():
+            pending_pairs = rejected_pair_collection.pop_pairs(3)
```
[the remark this answers is not in the document]
Comment (chen0040) [reply src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/dataset_generator.py]: updated based on comments

File: src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/dataset_generator.py [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+    return sql
+
+
+class SeedEvalDatasetGenerator:
+
+    def __init__(self):
+        self._genai_client = GenAiClient(timeout_secs=45)
+
+    async def generate_sql(
+        self,  task_working_dir: str) -> list[dict]:
```
[the remark this answers is not in the document]
Comment (chen0040) [reply src/google/cloud/db_context_enrichment/dataset/eval_seed_gen/dataset_generator.py]: updated based on comments

File: pyproject.toml [hunk 7]
```diff
@@ -1,14 +1,14 @@
 [project]
 name = "google-cloud-db-context-enrichment"
-version = "0.5.0"
```
Comment (wangauone) [inline pyproject.toml]: We don't need to bump version by ourselves. This will be handled by the release-please automatically.

File: golden_flight_booking_expert.json [hunk 8]
```diff
@@ -0,0 +1,32 @@
+[
+  {
+    "id": "flight_expert_001",
```
Comment (wangauone) [inline golden_flight_booking_expert.json:3]: IIUC, this and the following `golden_property_search_advanced.json` are part of the eval. We might not want to put these in the root folder.

File: .python-version [hunk 9]
```diff
@@ -1 +1 @@
-3.12
+3.12.13
```
Comment (wangauone) [inline .python-version:1]: Why this change?

File: tests/google/cloud/db_context_enrichment/main_test.py [hunk 10]
```diff
@@ -83,7 +83,7 @@ def test_attach_context_set_standard(tmp_path, clean_context_set_json):
     standard_file.write_text(json.dumps(standard_content))
 
     # Call the tool
-    attach_context_set.fn(
```
Comment (wangauone) [inline tests/google/cloud/db_context_enrichment/main_test.py]: I think these has been fixed in the main branch. We can do a quick merge of main to this PR?

File: skills/autoctx-dataset-generation/SKILL.md [hunk 11]
```diff
@@ -3,48 +3,79 @@ name: skill-autoctx-dataset-generation
...
 
-10. **Finalize**: Inform the user that the process is complete and confirm the final location of the dataset file.
+8.  **Prompt for Generation/Expansion & Complexity Tuning**: Ask the user if they want to generate new pairs. If yes, ask them to define the desired **SQL Complexity Level**:
+    *   *Level 1 (Simple)*: Basic filtering and sorting (`SELECT`, `WHERE`, `ORDER BY`, `LIMIT`).
+    *   *Level 2 (Intermediate)*: Aggregations and basic relationships (`GROUP BY`, `HAVING`, `INNER/LEFT JOIN`, Date/Math functions).
+    *   *Level 3 (Advanced)*: Complex logic (`CTEs`, Subqueries, Multiple `JOIN`s, `CASE WHEN` conditional logic).
+    *   *Level 4 (Expert)*: Analytical operations (Window functions like `RANK()`/`ROW_NUMBER()`, Self-joins, Pivot logic, handling complex JSON/Array data types).
 
-The standard evaluation format is a JSON object:
+9.  **Generate Dataset Pairs**:
```
Comment (wangauone) [inline skills/autoctx-dataset-generation/SKILL.md:52]: Checking the preview page, it seems there are some formatting issues that these are not showing as bullet points.

File: skills/autoctx-dataset-generation/SKILL.md [hunk 12]
```diff
@@ -3,48 +3,79 @@ name: skill-autoctx-dataset-generation
...
+3.  **Acquire Context, Resolve Schema, & Establish Semantic Bridge**: 
+    *   **Fetch Schema**: Attempt to use the `<source>-list-schemas` MCP tool.
+    *   **Schema Conflict Resolution (Source of Truth)**: If the user provided Business Context Artifacts or Application Source Code that contain schema definitions, you MUST cross-check the MCP tool's output against them. 
+        *   If the MCP tool returns an empty schema or a schema that fundamentally conflicts with the domain or tables described in the user's offline artifacts, treat the provided artifacts and source code as the single, infallible source of truth for the schema. Inform the user of this mismatch and your decision to use the offline schema.
+    *   **Synthesize Business Logic**: Map the business definitions from the documents to your finalized schema. 
+    *   **CRITICAL - Vocabulary Separation**: You must act as a semantic bridge. The **NLQ** must strictly use natural business terminology (e.g., "Active Users"). The **SQL** must strictly adhere to the technical schema (e.g., `WHERE status = 1`). Never invent column names in SQL based on business docs, and never leak raw column names into the NLQ.
```
Comment (wangauone) [inline skills/autoctx-dataset-generation/SKILL.md:24]: This is a nice one. I think we would like to make this as a hard rule

File: skills/autoctx-dataset-generation/SKILL.md [hunk 13]
```diff
@@ -3,48 +3,79 @@ name: skill-autoctx-dataset-generation
...
+    *   **If Query Logs are provided**: 
+        1. Extract the queries and rigorously filter out administrative or DML queries (`INSERT`, `UPDATE`, `DELETE`, `SELECT 1`, etc.). Keep only meaningful analytical `SELECT` statements.
+        2. Perform **Reverse Translation**: Translate the extracted SQL queries into natural language business questions. You MUST apply the *Semantic Bridge* rule here—the generated NLQs should sound like a business user asking a question, using the vocabulary from any provided business artifacts.
+        3. Treat these reverse-translated pairs as your "Seed Pairs" for the rest of the workflow.
 
-7.  **Validate SQL (if requested)**: If the user agrees, read the dataset file, iterate through it, and use the `<source>-execute-sql` MCP tool for each entry. Report any failures. Overwrite the file with any corrections if the user approves them.
+5.  **Initial Save (If Seed Pairs Provided or Extracted)**: If seed pairs were provided explicitly OR extracted/translated from Query Logs, use the `generate_dataset` MCP tool to save them. You must provide the exact `output_file_path`. Pass the constructed dataset as a JSON string (`dataset_entries_json`).
```
Comment (wangauone) [inline skills/autoctx-dataset-generation/SKILL.md:35]: How to use the tool would better be a tool description.

File: skills/autoctx-dataset-generation/SKILL.md [hunk 14]
```diff
@@ -3,48 +3,79 @@ name: skill-autoctx-dataset-generation
 description: "Generate and expand datasets of Natural Language Questions (NLQ) and SQL pairs for evaluation."
 ---
 
-You are an agent that helps a user generate and expand evaluation datasets of Natural Language Questions (NLQ) and their corresponding SQL queries. Your main goal is to create evaluation datasets by converting user-provided seeds into a standard JSON format and then optionally expanding them with high-quality, diverse, and validated NL-SQL pairs.
+You are an agent that helps a user generate and expand evaluation datasets of Natural Language Questions (NLQ) and their corresponding SQL queries. Your main goal is to create high-fidelity evaluation datasets by converting user-provided seeds into a standard JSON format, actively validating them for logical alignment, and expanding them using tunable complexity levels.
```
Comment (wangauone) [inline skills/autoctx-dataset-generation/SKILL.md:6]: Considering only put the high level key info in the skill when the content is getting bigger. We can put details to the skill references to reduce the context consumption here.

File: skills/autoctx-dataset-generation/SKILL.md [hunk 15]
```diff
@@ -3,48 +3,79 @@ name: skill-autoctx-dataset-generation
...
+    *   **Dialect & Aliasing Strictness**: Ensure the query perfectly conforms to the target database dialect (e.g., exact date functions). Ensure *every* column reference in a multi-table query is fully qualified with a table alias (e.g., `t1.id` instead of just `id`) to prevent ambiguous column errors.
+    *   **Ambiguity Detection**: Analyze the NLQ for vague terminology (e.g., "recent", "top", "best", "active"). If found, ensure the SQL explicitly resolves this. Does the reverse-translation of the SQL expose implicit assumptions not stated in the NLQ?
+    *   **Edge-Case Stress Test**: Consider how the query handles ties (e.g., `LIMIT 5` without tie-breakers), NULL values, and exact date boundaries. 
+    *   **Action**: Present any discovered flaws to the user and suggest specific corrections. Overwrite the file with user-approved corrections.
 
-10. **Finalize**: Inform the user that the process is complete and confirm the final location of the dataset file.
+8.  **Prompt for Generation/Expansion & Complexity Tuning**: Ask the user if they want to generate new pairs. If yes, ask them to define the desired **SQL Complexity Level**:
```
Comment (wangauone) [inline skills/autoctx-dataset-generation/SKILL.md:46]: Have we considered to separate the seed generation & dataset expansion to 2 skills?

cc: @helloeve @g-lynnzee

File: skills/autoctx-dataset-generation/SKILL.md (hunk 13 above)
Comment (wangauone) [inline skills/autoctx-dataset-generation/SKILL.md:35]: I forgot why we want to do initial save here. Wondering if we can save at the end of the workflow?

File: skills/autoctx-dataset-generation/SKILL.md [hunk 16]
```diff
@@ -3,48 +3,79 @@ name: skill-autoctx-dataset-generation
...
+        3. Treat these reverse-translated pairs as your "Seed Pairs" for the rest of the workflow.
 
-7.  **Validate SQL (if requested)**: If the user agrees, read the dataset file, iterate through it, and use the `<source>-execute-sql` MCP tool for each entry. Report any failures. Overwrite the file with any corrections if the user approves them.
+5.  **Initial Save (If Seed Pairs Provided or Extracted)**: If seed pairs were provided explicitly OR extracted/translated from Query Logs, use the `generate_dataset` MCP tool to save them. You must provide the exact `output_file_path`. Pass the constructed dataset as a JSON string (`dataset_entries_json`).
 
-8.  **Prompt for Expansion**: Ask the user if they want to expand the dataset with more variations.
+6.  **Prompt for Validation (If Seed Pairs Provided)**: If a seed was saved, ask the user if they want to validate the `golden_sql` and NLQ alignment in the dataset file. Advise them that this step ensures high-quality evaluation data.
```
Comment (wangauone) [inline skills/autoctx-dataset-generation/SKILL.md:37]: For validation, consider using the validation loop as mentioned in the best practices. 

https://agentskills.io/skill-creation/best-practices#validation-loops
